### PR TITLE
Added --keep-archive for GHES migrations

### PR DESCRIPTION
**Pull request by spindev**
Solved #831

- [x] Did you write/update appropriate tests
- [x] Release notes updated (if appropriate)
- [x] Appropriate logging output
- [x] Issue linked
- [x] Docs updated (or issue created)
- [x] New package licenses are added to `ThirdPartyNotices.txt` (if applicable)

### CODE REVIEW

**Comment by spindev** [comment]
#831

**Comment by timrogers** [comment]
@spindev Thanks for your contribution! We'll review this and get back to you.

File: src/OctoshiftCLI.Tests/gei/Handlers/MigrateRepoCommandHandlerTests.cs
Comment (ArinGhazarian) [inline src/OctoshiftCLI.Tests/gei/Handlers/MigrateRepoCommandHandlerTests.cs:1796]: nit: no need to assert these here since other tests are covering these. Just asserting that it doesn't delete should be sufficient enough.

**Review by ArinGhazarian — CHANGES_REQUESTED** [review]
Thanks for the contribution @spindev, 

Your implementation is good but with a few shortcoming that need to be addressed:
1. The `--keep-archive` option needs to be added to `gei generate-script` command too.
2. The `--keep-archvie` flag is only useful for migration from GHES before 3.8. So if it was provided for non GHES < 3.8 migrations, we would like to log a warning message and let the user know that there is no archive to keep basically. It can be added in the `ValidateGHESOptions` method.
3. You need to log the option if it's provided with its value (in our case it'd be `true`).

**Comment by timrogers** [comment]
> 3. You need to log the option if it's provided with its value (in our case it'd be `true`).

@spindev FYI, for #3, @ArinGhazarian is referring to a log statement like this one: https://github.com/github/gh-gei/blob/05d9dd5746c5eeda93fb8ef46b2457c3714c7b4f/src/gei/Handlers/MigrateRepoCommandHandler.cs#L402-L405.

File: src/OctoshiftCLI.Tests/gei/Handlers/MigrateRepoCommandHandlerTests.cs
Comment (spindev) [reply src/OctoshiftCLI.Tests/gei/Handlers/MigrateRepoCommandHandlerTests.cs]: Thx. Fixed.

**Comment by spindev** [comment]
Thanks for your feedback @ArinGhazarian and @timrogers. Review again please. Hopefully I got everything :)

File: src/gei/Commands/GenerateScriptCommand.cs
Comment (ArinGhazarian) [inline src/gei/Commands/GenerateScriptCommand.cs:40]: nit: remove the new line

File: src/gei/Handlers/GenerateScriptCommandHandler.cs
Comment (ArinGhazarian) [inline src/gei/Handlers/GenerateScriptCommandHandler.cs:177]: nit
```suggestion
            _log.LogInformation("KEEP ARCHIVE: true");
```

File: src/gei/Commands/MigrateRepoCommand.cs
Comment (ArinGhazarian) [inline src/gei/Commands/MigrateRepoCommand.cs:46]: nit: remove the new line

File: src/gei/Handlers/MigrateRepoCommandHandler.cs
Comment (ArinGhazarian) [inline src/gei/Handlers/MigrateRepoCommandHandler.cs:508]: 1. This is good 👍 not sure about the copy, @timrogers can help here 😃 
2. Other than this, another validation can also be added for when `GhesApiUrl` doesn't have a value but `--keep-archive` does. Basically in the else part of `if (args.GhesApiUrl.HasValue())`. We still probably shouldn't throw and just log a warning as you did here with a message like `--keep-archive can only be provided with --ghes-api-url.`

**Review by ArinGhazarian — CHANGES_REQUESTED** [review]
Looks good, just a few extra stuff.

File: src/gei/Handlers/MigrateRepoCommandHandler.cs
Comment (timrogers) [reply src/gei/Handlers/MigrateRepoCommandHandler.cs:508]:
```suggestion
                if (args.KeepArchive)
                {
                    _log.LogWarning("Ignoring --keep-archive option because there is no downloaded archive to keep");
                }
```

Comment (timrogers) [reply src/gei/Handlers/MigrateRepoCommandHandler.cs:508]: I think we could error in the case where the `--ghes-api-url` is not set and `--keep-archive` is set, because in this case, we know at the very beginning of the command that the combination of options is invalid, whereas in the above case, we only know once we've checked the GHES version.

**Comment by spindev** [comment]
@ArinGhazarian and @timrogers thanks for the feedback. There you go for the next review :)

File: src/gei/Commands/GenerateScriptCommand.cs
Comment (timrogers) [inline src/gei/Commands/GenerateScriptCommand.cs:117]: One more thing from me - but otherwise I'm good! This would also apply to `MigrateRepoCommand`:

```suggestion
            Description = "Keeps the archive on this machine after uploading to the blob storage account. Only applicable for migrations from GitHub Enterprise Server versions before 3.8.0."
```

**Review by timrogers — APPROVED** [review]

File: src/gei/Commands/GenerateScriptCommand.cs
Comment (spindev) [reply src/gei/Commands/GenerateScriptCommand.cs]: Done :)

File: src/gei/Handlers/MigrateRepoCommandHandler.cs
Comment (ArinGhazarian) [inline src/gei/Handlers/MigrateRepoCommandHandler.cs:593]: A unit test needs to be added for this since we're throwing here.

**Review by ArinGhazarian — CHANGES_REQUESTED** [review]
It looks great 🎉 just add a unit test for the new validation that throws and this should be good to go.

**Review by ArinGhazarian — APPROVED** [review]
🎉

**Comment by spindev** [comment]
@timrogers and @ArinGhazarian Thank you guys. Do I need to do additional stuff before the PR could be merged?

**Comment by timrogers** [comment]
I think this is good to go, but I'd like to get a review from an engineer before we merge.

**Comment by dylan-smith** [comment]
Can you add a bullet point to RELEASENOTES.md with this text:
`- Added `--keep-archive` flag to `gh gei migrate-repo` and `gh gei generate-script`. When migrating from GHES this will skip the step where we delete the archive from your machine, leaving it around as a local file.`

I tried to do it but I don't have permissions to push to your branch/fork. Once that's in there I'll merge this in.

Thanks!

**Review by dylan-smith — CHANGES_REQUESTED** [review]
looks good, just needs a line in release notes and it's good to merge

**Comment by spindev** [comment]
> looks good, just needs a line in release notes and it's good to merge

Done! :)

**Review by dylan-smith — APPROVED** [review]

**Pull request merged**
